src/librbd/AioRequest.cc

```diff
@@ -52,37 +57,186 @@ namespace librbd {
-			    m_object_no, m_object_off, m_object_len,
-			    image_extents);
+    switch (m_state) {
+    case LIBRBD_AIO_READ_GUARD:
+      ldout(m_ictx->cct, 20) << "should_complete " << this
+                             << " READ_CHECK_GUARD" << dendl;
+
+      // This is the step to read from parent
+      if (!m_tried_parent && r == -ENOENT) {
+        RWLock::RLocker l(m_ictx->snap_lock);
+        RWLock::RLocker l2(m_ictx->parent_lock);
+
+        if (m_ictx->parent == NULL) {
+	  ldout(m_ictx->cct, 20) << "parent is gone; do nothing" << dendl;
+	  m_state = LIBRBD_AIO_READ_FLAT;
+	  finished = false;
+	  break;
+	}
 
-      uint64_t image_overlap = 0;
-      r = m_ictx->get_parent_overlap(m_snap_id, &image_overlap);
-      if (r < 0) {
-	assert(0 == "FIXME");
+        // calculate reverse mapping onto the image
+        vector<pair<uint64_t,uint64_t> > image_extents;
+        Striper::extent_to_file(m_ictx->cct, &m_ictx->layout,
+			        m_object_no, m_object_off, m_object_len,
+			        image_extents);
+
+        uint64_t image_overlap = 0;
+        r = m_ictx->get_parent_overlap(m_snap_id, &image_overlap);
+        if (r < 0) {
+          assert(0 == "FIXME");
+        }
+        uint64_t object_overlap = m_ictx->prune_parent_extents(image_extents, image_overlap);
+        if (object_overlap) {
+          m_tried_parent = true;
+
+          if (is_copy_on_read(m_ictx, m_snap_id)) {
+            // If there is a valid object being coyping up, directly extract
+            // content and finish up.
+            Mutex::Locker l3(m_ictx->copyup_list_lock);
+
+            map<uint64_t, CopyupRequest*>::iterator it =
+              m_ictx->copyup_list.find(m_object_no);
+            if (it != m_ictx->copyup_list.end()) {
+              Mutex::Locker l4(it->second->get_lock());
+              if (it->second->is_ready()) {
+                ceph::bufferlist &copyup_data = it->second->get_copyup_data();
+                m_read_data.substr_of(copyup_data, m_object_off, m_object_len);
+                ldout(m_ictx->cct, 20) << __func__ << " extract content from copyup_list, obj-"
+                                       << m_object_no << dendl;
+                finished = true;
+                break;
+              }
+            }
+          }
+
+          read_from_parent(image_extents);
+
+          if (is_copy_on_read(m_ictx, m_snap_id))
+            m_state = LIBRBD_AIO_READ_COPYUP; 
+          else 
+            m_state = LIBRBD_AIO_READ_GUARD;
+
+          finished = false;
+        }
+      }
+      break;
+    case LIBRBD_AIO_READ_COPYUP:
+      ldout(m_ictx->cct, 20) << "should_complete " << this << " READ_COPYUP" << dendl;
+      // This is the extra step for copy-on-read: kick off an asynchronous copyup.
+      // It is different from copy-on-write as asynchronous copyup will finish
+      // by itself so state won't go back to LIBRBD_AIO_READ_GUARD.
+
+      if (r > 0) {
+        assert(m_tried_parent);
```

Comment:
You can move the ```assert``` in front of the ```if``` statement since you should never hit this state w/o going through ```LIBRBD_AIO_READ_GUARD```